flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/operations/utils/factories/CalculatedTableFactory.java

```diff
@@ -95,42 +94,87 @@ public FunctionTableCallVisitor(String[] leftTableFieldNames) {
 			}
 
 			CallExpression tableCall = (CallExpression) children.get(0);
-			TableFunctionDefinition tableFunctionDefinition =
-				(TableFunctionDefinition) tableCall.getFunctionDefinition();
-			return createFunctionCall(tableFunctionDefinition, aliases, tableCall.getResolvedChildren());
+			return createFunctionCall(tableCall, aliases);
+		}
+
+		private CalculatedQueryOperation<?> createFunctionCall(CallExpression call, List<String> aliases) {
+			FunctionDefinition definition = call.getFunctionDefinition();
+			if (definition instanceof TableFunctionDefinition) {
+				return createFunctionCall(
+					((TableFunctionDefinition) definition).getTableFunction(),
+					call.getFunctionIdentifier().orElse(FunctionIdentifier.of(definition.toString())),
+					call.getOutputDataType(),
+					aliases,
+					call.getResolvedChildren());
+			} else if (definition instanceof TableFunction<?>) {
+				return createFunctionCall(
+					(TableFunction<?>) definition,
+					call.getFunctionIdentifier().orElse(FunctionIdentifier.of(definition.toString())),
+					call.getOutputDataType(),
+					aliases,
+					call.getResolvedChildren());
+			} else {
+				return defaultMethod(call);
+			}
 		}
 
 		private CalculatedQueryOperation<?> createFunctionCall(
-				TableFunctionDefinition tableFunctionDefinition,
+				TableFunction<?> tableFunction,
+				FunctionIdentifier identifier,
+				DataType resultType,
 				List<String> aliases,
 				List<ResolvedExpression> parameters) {
-			TypeInformation<?> resultType = tableFunctionDefinition.getResultType();
 
-			int callArity = resultType.getTotalFields();
-			int aliasesSize = aliases.size();
+			final TableSchema tableSchema = adjustNames(
+				extractSchema(resultType),
+				aliases,
+				identifier);
 
-			String[] fieldNames;
+			return new CalculatedQueryOperation(
+				tableFunction,
```

Comment:
Pass the function definition as a marker for `QueryOperationConverter`. Otherwise it will still use the old stack. We should use `BridgingSqlFunction` there. We should be able to use `relBuilder.functionScan`.